Adding verification of admin email into admincp settings.

WhiteNeo opened the pull request:
About this; https://github.com/mybb/mybb/issues/3955#issuecomment-619433817

Only i need to add it into update file so i need to verify that too, to add it automatically if empty.

Resolves #3955

WhiteNeo committed:
Adding verification of admin email into admincp settings.

About this; https://github.com/mybb/mybb/issues/3955#issuecomment-619433817

Only i need to add it into update file so i need to verify that too, to add it automatically if empty.

WhiteNeo committed:
Update upgrade50.php

Adding admin email verification. (Need to verify i am not sure if it works but it have to).

euantorano commented on inc/languages/english/admin/config_settings.lang.php:
```diff
@@ -90,6 +90,7 @@
 $l['error_ajax_search'] = "There was a problem searching for settings:";
 $l['error_ajax_unknown'] = "An unknown error occurred while searching for settings.";
 $l['error_chmod_settings_file'] = "The settings file \"./inc/settings.php\" isn't writable. Please CHMOD to 666.<br />For more information on CHMODing, see the <a href=\"https://docs.mybb.com/1.8/administration/security/file-permissions\" target=\"_blank\" rel=\"noopener\">MyBB Docs</a>.";
+$l['error_admin_email_settings_empty'] = "Admin Email can not be empty, and have to be a valid email address";
```
```suggestion
$l['error_admin_email_settings_empty'] = "Admin Email can not be empty, and must be a valid email address";
```

euantorano commented on install/resources/upgrade50.php:
```diff
@@ -35,3 +35,34 @@ function upgrade50_dbchanges()
+	$output->print_header("Verifying Admin Email");
+	if(empty($mybb->settings['adminemail']))
+	{
+		echo "<p>Updating admin email into settings...</p>";
```
```suggestion
		echo "<p>Updating admin email setting...</p>";
```

euantorano commented on install/resources/upgrade50.php:
```diff
@@ -35,3 +35,34 @@ function upgrade50_dbchanges()
+				$email = $db->fetch_field($query, 'email');
+				$db->update_query('settings', array('value' => $email), "name='adminemail'");
+			}
+		}
```
It looks like there's a case where the email may still be empty after this - I'm not sure how we want to handle that..?

WhiteNeo replied:
I have seen multiple boards where admin accounts are with empty emails due maybe db manipulation.

I think that is unnecessary anyways but i add only a case just to add a valid email if current user who runs the update don't have that field and go throught the update without any other issue. I think that part is not necessary at all. And with only run the update must be enought to add current user email :) Im that file as i try to set a comment yeah it needs a deep review from you due the people who knows more about the mybb system is the staff i think. But yeah this was a sugestion only on the end to try to help you in some way because you have given to much to all of us and is what i think we have to do at least, try to contribute.

Sama34 replied:
I think @euantorano means that administrators can empty the setting after updating. This shouldn't be allowed.

yuliu replied:
> I think @euantorano means that administrators can empty the setting after updating. This shouldn't be allowed.

No I don't think so.

This query `$query = $db->simple_select('users','email',"usergroup=4 AND email<>''", array("limit" => 1,"order_by" => "id","order_dir" => "asc"));` may return no result, so the `adminemail` will still be empty.

euantorano reviewed (changes requested):
Some minor language suggestions, and a comment about the upgrade process that may need some discussion.

JordanMussi commented on install/resources/upgrade50.php:
```diff
@@ -35,3 +35,34 @@ function upgrade50_dbchanges()
 	$output->print_contents("<p>Click next to continue with the upgrade process.</p>");
 	$output->print_footer("50_done");
```
```suggestion
	$output->print_footer("50_verify_adminemail");
```

JordanMussi replied:
You've marked this comment as resolved @WhiteNeo but not made the changes? There should be a commit suggestion below the comment that will make the change for you. 😄

WhiteNeo replied:
I see, i have used edit button and changed the files but i was thinked on press it takes automatically your sugestions, that why i have pressed the button 😂

WhiteNeo committed:
Update config_settings.lang.php

WhiteNeo committed:
Update upgrade50.php

Resolving the current review to let this solved as fast as we can :)

WhiteNeo commented:
I have updated the files into my repo, is that fine ?

Or i have to do domething else, please let me know and so many thanks for reviewing it. This is the less i can do for this great project.

WhiteNeo reviewed (commented):
Files updated and reviewed, thanks for the recomendations.

yuliu commented:
Had a comment at https://github.com/mybb/mybb/issues/3955#issuecomment-626783266 .

Using the admin's email who executed the upgrade script seems better than fetching the first email of users in admin groups, though I still doubt its legitimacy.

If we can't find an applicable way, @Sama34 's idea of an upgrade notice is the last resort: https://github.com/mybb/mybb/issues/3955#issuecomment-619461167

WhiteNeo commented:
The code takes the user who update the script however sometimes that field is empty into some boards for some weird reason and i use the first admin usergroup user by default who have an email. However that was unnecessary because as far as i can remember that is a required foeld, so i have updated the files and the pr seems to be up to date. But i have no idea if i have done in right way because i have never used before github and this is a new challenge for me. Only i have updated files and uploaded it to my repositories. But i have no idea on how this works :)

I have updated the files, do i need to sent again or how can i do to update the changed files i have reviewed ?

yuliu commented:
> The code takes the user who update the script however sometimes that field is empty into some boards for some weird reason and i use the first admin usergroup user by default who have an email. However that was unnecessary because as far as i can remember that is a required foeld, so i have updated the files and the pr seems to be up to date.

Yeah from your updated PR I get that idea. I just have the doubt about if using an email address of a user, even if the user is an admin or anyone who executed the upgrade script, is legitimate.

>But i have no idea if i have done in right way because i have never used before github and this is a new challenge for me. Only i have updated files and uploaded it to my repositories. But i have no idea on how this works :)
> I have updated the files, do i need to sent again or how can i do to update the changed files i have reviewed ?

You're doing great work! Thanks. @JordanMussi means that you haven't updated the upgrade script about `$output->print_footer("50_done");`, see here: https://github.com/mybb/mybb/pull/3993#discussion_r423116079

WhiteNeo commented:
Yes indeed the file was changed automatically i think. I have updated it into the last file sent but i am not sure if it's fine on how i do.

But it's in here i think.

https://github.com/mybb/mybb/pull/3993#commits-pushed-ebf89a7

Sama34 commented:
> If we can't find an applicable way, @Sama34 's idea of an upgrade notice is the last resort: #3955 (comment)

> i use the first admin usergroup user by default who have an email.

This is what I think we should avoid. The email of each user is meant to be private, and thus the script shouldn't be  _randomly_ pickingany their emails for fallback.

WhiteNeo commented:
If you are an admin and are runing the update the information would not be showed due it adds it into admincp, however all that information is available if you go throught admincp on every user account and even that field as plain texts.

I think there must be another reason but anyways it's only to keep it as non empty and already will take the user email who run the upgrade and it's all.

Even any other account don't be shown or anything else and yeah it's private and would be as is, if you are an admin it's the same if you run or not the update some email must be taken and the advertisement let you know about that for you can setup the right one from now :) only when you are on that settings part of course.

Sama34 commented:
It doesn't matter if you can access the user information, using their emails is entirely out of your scope as a site administrator as you don't own the email account simply by knowing it.

This might not be a big deal as not just anyone is an administrator in any forum, yes. But even so I think we should avoid such route and simple force the setting to never be left empty and, if somehow emptied, fallback to other email setting, not users' emails.

Simply put: I don't agree with the reasoning for mixing board settings with user data.

yuliu commented:
> If we go this route then we should, for example, add a user setting to disable the system from picking their emails (for privacy or so), that is, honestly, totally unnecessary for fixing this.

It's not only a privacy concern but also a security concern. We shouldn't trust any email as `adminemail` unless the forum owner would input/specify it.

If a random mail address is chosen without the forum owner's consent, even though chosen from emails of known administrators but neither a desired one nor a secure one, as the `adminemail`, users who receive mails from this address would believe it's a valid and safe communication channel without hesitate and probably reply to this address with some sensitive information. That's just one foreseeable vulnerability.

WhiteNeo commented:
The reason why i use that way is because i trust in anyone that has a part of my forums with admin range if you don't think it was a good idea fine that information can be changed manually.

And i don't see a point of discusion due on the end i think you have not reviewed the end code lines, only are talking for what reason ?

On the end the final idea is to contribute but if you take it that way i think it's better to never more try to do it if you take as an offense or i don't know. I will not waste my time with nonsense things.

yuliu commented:
> The reason why i use that way is because i trust in anyone that has a part of my forums with admin range if you don't think it was a good idea fine that information can be changed manually.

Honestly, the way you trust every admins on your forum probably won't apply to all other forums. And MyBB neither does in this way, I think.

You'll get the same underlying concept from the design that each admin can have different AdminCP permission settings and maybe not all admins will be granted super admin privilege (designated manually by adding admin's ID in `inc/config.php` file).

And further, we should take extra care about altering settings for existing installs, since it may affect things in an unexpected way.

> And i don't see a point of discusion due on the end i think you have not reviewed the end code lines, only are talking for what reason ?

The point of this argument here is not to invalid the origin of this PR which specified in #3955, and also not to mark your PR as worthless. The following code does fix the hole that `adminemail` can be empty in AdminCP settings:
https://github.com/mybb/mybb/pull/3993/files#diff-d8ba16aa014c15e1fde4896a00d88005R1028-R1034 
And this part of code will get merged if this PR is approved, I think. Is that correct, @euantorano ?

And the point of discussions are about whether or not and how would we do with potential empty `adminemail` in exiting installs:
https://github.com/mybb/mybb/pull/3993/files#diff-1da9a318f936a17cf4b748d0c8043296R48
And on this point are we having discussions.

> On the end the final idea is to contribute but if you take it that way i think it's better to never more try to do it if you take as an offense or i don't know. I will not waste my time with nonsense things.

I'm so happy to see your PR and contribution is welcome and honored here. But I'm so sorry that you feel that way.

euantorano commented:
I’m totally happy with the code in the ACP. 

The installer/upgrade code I’m not so sure about. I did mean that the query can result in no rows, which would cause an empty field still. 

I wonder how feasible it would be to add an input box in the upgrade script for admins to input the email address to use. I can’t remember if that’s even possible with the current upgrade system code. 

I’m on my phone now, but don’t we have a setting for the authentication username for SMTP? Perhaps we could use that value if it’s an email address. 

On Tue, 12 May 2020, at 07:37, yuliu wrote:
> 
> 
> *@yuliu* commented on this pull request.
> 
> In install/resources/upgrade50.php 
> <https://github.com/mybb/mybb/pull/3993#discussion_r423494100>:
> 
> > +	{
> +		echo "<p>Updating admin email into settings...</p>";
> +		echo "<p><small>P.D: Field can not be empty</small></p>";
> +		if($mybb->user['email'] != "")
> +		{
> +			$db->update_query('settings', array('value' => 
> $mybb->user['email']), "name='adminemail'");
> +		}
> +		else
> +		{
> +			$query = $db->simple_select('users','email',"usergroup=4 AND 
> email<>''", array("limit" => 1,"order_by" => "id","order_dir" => 
> "asc"));
> +			if($db->num_rows($query) == 1)
> +			{
> +				$email = $db->fetch_field($query, 'email');
> +				$db->update_query('settings', array('value' => $email), 
> "name='adminemail'");
> +			}
> +		}
> > I think @euantorano <https://github.com/euantorano> means that administrators can empty the setting after updating. This shouldn't be allowed.
> 
> No I don't think so.
> 
> This query `$query = $db->simple_select('users','email',"usergroup=4 
> AND email<>''", array("limit" => 1,"order_by" => "id","order_dir" => 
> "asc"));` may return no result, so the `adminemail` will still be empty.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mybb/mybb/pull/3993#discussion_r423494100>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAFW24NDJW4JELXT7UKDXCDRRDVDJANCNFSM4M543UBQ>.
>

yuliu commented:
> The installer/upgrade code I’m not so sure about.

When comes to the install script, it does check if `adminemail` is empty but that's only. IMO this check during install could be improved in this PR for checking the email address's validity. See https://github.com/mybb/mybb/blob/mybb_1822/install/index.php#L2208-L2211 and https://github.com/mybb/mybb/blob/mybb_1822/install/resources/language.lang.php#L331

> I wonder how feasible it would be to add an input box in the upgrade script for admins to input the email address to use. I can’t remember if that’s even possible with the current upgrade system code.

I also had this thought of adding an input box during upgrade but don't know if it's applicable. Was thinking about the [auto-submit](https://github.com/mybb/mybb/issues/3841), but that's for older upgrades. Now it seems possible to me. I think we may give it a try.

However, if 1.9's upgrade uses 1.8's schemes that for older upgrade scripts the page will submit itself, this way may have its dead end.

> I’m on my phone now, but don’t we have a setting for the authentication username for SMTP? Perhaps we could use that value if it’s an email address.

Yes it is, for forums that have SMTP configured but it could also result in no `adminemail` added if SMTP isn't configured or a username rather than an email address is used for authentication. Anyway, the chosen email address is somehow still undesirable to me.

yuliu commented:
As of now, there are several ways of "setting a valid admin email for forums whose such value is empty" discussed in this PR:
- In the upgrade process:
  1. Use the email of the user who upgrades the forum.
  1. Use the email of the user who is an administrator and has a smaller UID, if any.
  1. Use the authentication username for SMTP, if it exists and it's an email address.
  1. Set up an input box during upgrading.
- Before / After the upgrade process:
  1. Issue an upgrade warning in release post.
  1. Add some alert functions to AdminCP so that forum administrators can be alerted by errors/warnings about invalid/empty settings, during AdminCP login or in a specific header area.

euantorano commented:
> As of now, there are several ways of "setting a valid admin email for forums whose such value is empty" discussed in this PR:
> 
> * In the upgrade process:
>   
>   1. Use the email of the user who upgrades the forum.
>   2. Use the email of the user who is an administrator and has a smaller UID, if any.
>   3. Use the authentication username for SMTP, if it exists and it's an email address.
>   4. Set up an input box during upgrading.
> * Before / After the upgrade process:
>   
>   1. Issue an upgrade warning in release post.
>   2. Add some alert functions to AdminCP so that forum administrators can be alerted by errors/warnings about invalid/empty settings, during AdminCP login or in a specific header area.

My preference would be a combination of some of the above, in this order:

1. Use the authentication username for SMTP, if it exists and it's an email address.
2. Use the email of the user who upgrades the forum if it is set and is a valid email.
3. Set up an input box in the upgrader.

I think we should definitely include a note in the release post no matter what.

I think we should introduce an ACP sanity check no matter what, which will cover key required settings. This should appear as a banner at the top of all ACP pages if the check fails.

yuliu commented:
> My preference would be a combination of some of the above, in this order:
>     1. Use the authentication username for SMTP, if it exists and it's an email address.
>     2. Use the email of the user who upgrades the forum if it is set and is a valid email.
>     3. Set up an input box in the upgrader.

I'm not sure but the mail address used for SMTP authentication might be a non-reply one or a mailbox dedicated to sending mails rather than receiving mails.

Anyway, I suggest that an alert message should be given on the page to notify the admin user about what the `adminemail` is now if it's been changed, or needing to manually change it in AdminCP after upgrading. For example:
```
The board's <strong>Admin Email</strong> has been change to XXX@YYY.ZZZ, please review it in the Admin Control Panel after upgrade.<br />
The setting of <strong>Admin Email</strong> is located at <em>Admin Control Panel » Configuration » Board Settings » Site Details</em>.
```
and
```
The board's <strong>Admin Email</strong> is missing or is an invalid email address, please review and enter a valid one for it in the Admin Control Panel after upgrade.
The setting of <strong>Admin Email</strong> is located at <em>Admin Control Panel » Configuration » Board Settings » Site Details</em>.
```

Sama34 commented:
> And i don't see a point of discussion due on the end i think you have not reviewed the end code lines, only are talking for what reason ?

Creo que estas confundiendo nuestra intención. Es claro que tu solución puede funcionar para tu comunidad al igual que puede funcionar para mucha gente (incluido yo), pero no es una solución que funciona para todos y por tal motivo no se puede forzar tal cual.

Basicamente, nuestra preocupaciones son 2:
- Primero, el sistema no debe de arbitrariamente utilizar la información privada de usuarios al azar y sin limites. Si tus administradores están de acuerdo en utilizar sus correos para el sistema de mail entonces tan sencillo como colocar esos correos en las opciones globales del foro, no hay necesidad de buscarlos entre datos de usuarios.
- Como @yuliu menciona, no solo es un asunto de privacidad (la cual podríamos ignorar, como en tu caso que funciona bien así), pero también se trata del usuario que recibe el correo ya que este podría confundir la información (el correo) y considerarla como un dato de comunicación segura, cuando posiblemente no lo es. (ejemplo: recibo un email automatico de ventas.com pero el correo que me muestra es juanito@gmail.com ..? Que tanto puedo confiar en esto como usuario [no como administrador]?)

> On the end the final idea is to contribute but if you take it that way i think it's better to never more try to do it if you take as an offense or i don't know. I will not waste my time with nonsense things.

Se agradece tu contribucion al igual que la de todos, e incluso si no se acepta se agradece el esfuerzo y la discusión a la que contribuyes. Sin embargo, en este caso específico, lamento decir que elegiste un problema muy complejo en el cual contribuir y esperar que tu solución sea aceptada sin discutir. Se está discutiendo para encontrar la mejor estrategia que solucione y funcione para todos.

Quizás ahora entiendas (y otros) por qué algunos problemas tardan en solucionarse. No siempre se puede estar 100% de acuerdo en todo, incluso si eres o no parte del equipo oficial.

Así como hay PR que se aceptan prácticamente sin ningún comentario también hay algunos que tardan mucho y en ocasiones no se aceptan tal cual se enviaron.

Anyways, I would expect this to be fixed for 1.8.23 so your contribution isn't just going to be wasted, if that is what you feel.

> And the point of discussions are about whether or not and how would we do with potential empty `adminemail` in exiting installs:

Maybe we can simply ignore this and add a check for all boards that display a warning in the ACP that points to the docs where it explains why the setting shouldn't be empty? I see this working for all and we won't need to add a code that forces this setting to not be saved empty nor any update code whatsoever.

> Add some alert functions to AdminCP so that forum administrators can be alerted by errors/warnings about invalid/empty settings, during AdminCP login or in a specific header area.

>I think we should introduce an ACP sanity check no matter what, which will cover key required settings. This should appear as a banner at the top of all ACP pages if the check fails.

Pretty much what you describe here.

> I wonder how feasible it would be to add an input box in the upgrade script for admins to input the email address to use.

This sounds like a useful enhacements for situations where new settings are added. But with the current code it would require quite some modifications. Maybe we can do this later on whenever we work on updating the upgrade script.

> Use the authentication username for SMTP, if it exists and it's an email address.

This would cause issues with some installations. I, for instance, use STMP but the username I use is not the same I use for all other settings here mentioned. Thus, it will make public a data that I' rather keep secret.

WhiteNeo commented:
@Sama34 la cuestión en si es la siguiente:

Ese trozo de código del que mencionan desde mucho antes fue removido.

https://github.com/mybb/mybb/pull/3993#commits-pushed-ebf89a7

Ahí dejé los cambios sugeridos desde la primer instancia, por eso se me hace extraño que hablen de un código que ya ni siquiera figura el el PR que actualice, y que supuestamente se realizo el cambio correctamente.

Pero yo no lo elegí ni mucho menos, Ben hizo la petición en discord para apoyar con algunos asuntos que eran primeramente este, así que intenté hacerlo en eso que se me pidió porque me gusta contribuir siempre en lo que me es posible, no sé y ni tengo idea de como funcione github realmente porque sólo lo usé cuando empezaba a realizar algunas cosas, pero desde que Microsoft tomó el control o lo iba a tomar yo decidí como muchos de sus productos que dejan de ser lo que eran abandonarlo de tajo. Yo me refiero más bien a esa parte que no le veo punto discutir sobre un aparte de código que ya ni siquiera existe en el PR desde mucho antes de sus comentarios. Pero ni tengo idea de como se maneja github porque aparte de eso me tuve que pedir el re envio de mis datos, descargarme los archivos nuevos para poder hacer los cambios, pensar en que podría servir a la causa de esta petición que dejé, pero sin idea siquiera de como hacerlo, simplemente dejé que el instinto me guiará y creí que hice lo adecuado, ahora bien les solicité me ayudaran para realizar bien las cosas y como no vi mensaje de apoyo alguno pensé que todo estaba correcto, por lo cual no tengo ni la más remota idea por eso dejo la liga de los últimos archivos revisados que dejé y simplemente dejé el código de el correo del usuario que actualiza en ese momento, cosa que tu mismo mencionaste estaría mejor, pero es lo que se hizo y en el código simplemente agregué más posibilidades que creí podrían ayudar, luego de sus comentarios me di cuenta que no era adecuado, sin embargo los cambios de esos trozos de código ya ni siquiera existían :) Por eso reitero que no le veo punto de discusión alguna a algo que no existe, debería enfocarse uno a ver como lo hace Euan Torano a lo que hace falta, me iba a poner un rato a intentar agregar alguna cosa que se me ocurriera, pero he sentido como que realmente podría no valer la pena y actualmente tengo varios proyectos pendientes en los cuales estoy trabajando y me dedicaré de lleno a ellos para no meter lío en otros lados donde no me llaman (es retórico, pues si me llamaron, intenté acudir al llamado y hacer lo que creí era buena idea), pero no lo malentiendo, tal vez sea la forma de escribir que uno intenta sea similar a nuestro idioma en contexto que a veces se tergiversa y es cuando hay mucha confusión, por eso ya mejor dejé de participar porque en realidad esos puntos que se han discutido creo que son innecesarios porque el código lo dejé en esa parte tal cual, de lo otro leí el mensaje que se me solicitó de apoyo y como es mi primer participación, simplemente creí que era algo sencillo y que a como yo lo creí era (Dar una idea plasmada del código y no sólo palabras, de ese código ustedes lo modificarían y dejarían al gusto y necesidades de la plataforma, pues yo desconozco el modus operandi de mybb, pero no pensé que sería como algo que uno quisiera fuese aprobado, simplemente era una idea posible de solucionar el problema actual de acuerdo a lo que ya habían ustedes hablado en los hilos del tema compartido y de ahí ya lo demás se dio sólo). Gracias a todos por sus comentarios, pero realmente yo sólo quería al final compartir mi idea que es como creí que esto de los PR funcionaba, y que finalmente ustedes ya cogían trozos de código y modificaban y ya está.

Por demás escribo bastante muchas veces y no llego a un punto concreto, pero espero no haber incomodado a la comunidad, en lugar de haber apoyado realmente, eso es todo.

Sama34 commented on inc/languages/english/admin/config_settings.lang.php:
```diff
@@ -90,6 +90,7 @@
 $l['error_ajax_search'] = "There was a problem searching for settings:";
 $l['error_ajax_unknown'] = "An unknown error occurred while searching for settings.";
 $l['error_chmod_settings_file'] = "The settings file \"./inc/settings.php\" isn't writable. Please CHMOD to 666.<br />For more information on CHMODing, see the <a href=\"https://docs.mybb.com/1.8/administration/security/file-permissions\" target=\"_blank\" rel=\"noopener\">MyBB Docs</a>.";
+$l['error_admin_email_settings_empty'] = "Admin Email can not be empty, and must be a valid email address";
```
```suggestion
$l['error_admin_email_settings_empty'] = "Admin Email can not be empty, and must be a valid email address.";
```

Sama34 commented:
Oh, bueno, quizás tu creas que un PR funciona de esa manera, en la cual tu envias una modificación y luego se toman las partes que se consideren bien recibidas y se deseche el resto. En realidad no creo que ese sea la forma en que funciona, ya que se espera que un PR solucione todo el problema de tajo o alguna parte concreta de un problema (cuando podría haber una diferencia notoria entre las distintas causas de un mismo problema).

En este caso como te menciono, decidiste apoyar con un problema que es complejo no por que la solución sea compleja, si no mas bien por que el como deberia ser la solución correcta es aun un tema que se está discutiendo (por eso la discusión que para ti parece no tener sentido).

Pero entiendo tu frustración, deja envió algunos comentarios para ver si eso funciona para todos y así damos por cerrado esto lo mas pronto posible...

Sama34 commented on admin/modules/config/settings.php:
```diff
@@ -1025,6 +1025,13 @@
+		// Verify for admin email that can't be empty
+		if(isset($mybb->input['upsetting']['adminemail']) && !filter_var($mybb->input['upsetting']['adminemail'], FILTER_VALIDATE_EMAIL))
+		{
+			flash_message($lang->error_admin_email_settings_empty, "error");
+			admin_redirect("index.php?module=config-settings&action=change&gid=".$gid);
```
```suggestion

			unset($mybb->input['upsetting']['adminemail']);
			$lang->success_settings_updated .= $lang->error_admin_email_settings_empty;
```

Sama34 replied:
@JordanMussi what about this? This looks inconsistent with how we handle the login methods and emails settings.

![image](https://user-images.githubusercontent.com/1786584/85985109-12870a80-b99f-11ea-9680-cf5031c1768f.png)

Sama34 replied:
Please note that the current approach ignores form changes entirely when redirecting the user to the same page.

Sama34 commented on install/resources/upgrade50.php:
```diff
@@ -35,3 +35,22 @@ function upgrade50_dbchanges()
+		echo "<p>Admin email verified success...</p>";		
+	}
+	$output->print_contents("<p>Click next to continue with the upgrade process.</p>");
+	$output->print_footer("50_done");
```
```suggestion
	$output->print_footer("50_update_adminemail");
```

<-- then change this to `$output->print_footer("50_update_adminemail");` and add a `upgrade50_update_adminemail`, something along the lines of... -->

Sama34 commented on install/resources/upgrade50.php:
```diff
@@ -35,3 +35,22 @@ function upgrade50_dbchanges()
+	{
+		echo "<p>Updating admin email settings...</p>";
+		echo "<p><small>P.D: Field can not be empty</small></p>";
+		$db->update_query('settings', array('value' => $mybb->user['email']), "name='adminemail'");
```
Maybe you can add a simple input field for users to insert an email here? ... ->
```suggestion
		if(!isset($mybb->input['adminemail']))
		{
			$mybb->input['adminemail'] = $mybb->user['email'];
		}
		echo "<input type=\"text\" class=\"text_input\" name=\"adminemail\" id=\"adminemail\" value=\"".htmlspecialchars_uni($mybb->input['adminemail'])."\" />";
```

Sama34 commented on install/resources/upgrade50.php:
```diff
@@ -35,3 +35,22 @@ function upgrade50_dbchanges()
+		echo "<p><small>P.D: Field can not be empty</small></p>";
+		$db->update_query('settings', array('value' => $mybb->user['email']), "name='adminemail'");
+	}
+	else
+	{
+		echo "<p>Admin email verified success...</p>";		
+	}
```
```suggestion

```

Sama34 commented on install/resources/upgrade50.php:
```diff
@@ -35,3 +35,22 @@ function upgrade50_dbchanges()
+	}
+	$output->print_contents("<p>Click next to continue with the upgrade process.</p>");
+	$output->print_footer("50_done");
+}
```
```suggestion
}

function upgrade50_update_adminemail()
{
	global $output, $cache, $db, $mybb;

	$output->print_header("Updating Admin Email");
	if(isset($mybb->input['adminemail']) && filter_var($mybb->input['adminemail'], FILTER_VALIDATE_EMAIL))
	{
		$db->update_query('settings', array('value' => $db->escape_string($mybb->input['adminemail'])), "name='adminemail'");
	}
	else
	{
		// I'm unsure if this would work to go back if user input is invalid..
		$output->print_contents("<p>Admin Email can not be empty, and must be a valid email address.</p>");
		$output->print_footer("50_verify_adminemail");
	}

	echo "<p>Admin email verified success...</p>";

	$output->print_contents("<p>Click next to continue with the upgrade process.</p>");
	$output->print_footer("50_done");
}
```

@euantorano @yuliu , what do you think? Does this seems like something that makes sense and could be enough?
```

WhiteNeo committed:
Update upgrade50.php

Updating to verify admin email address.

WhiteNeo committed:
Update upgrade50.php

Cleaning some parts of code and fix a redirection.

JordanMussi commented:
@Sama34 it would be better to use the same method as the main installer - it just calls the function for the previous step if there are errors:

https://github.com/mybb/mybb/blob/76bd4568a4cbbf35b6b3922b8605e40809d7d6d2/install/index.php#L1979-L1990

WhiteNeo committed:
Update upgrade50.php

Beterize of the code due Jordan Musi sugestion, this looks more usefull and to be the best approach for the final PR

WhiteNeo committed:
Update config_settings.lang.php

Add Omar G sugestion to make it propertly on lang change sugestion.

WhiteNeo reviewed (commented):
This are the last changes for this sugested PR

JordanMussi commented:
Just double checked this and all seems to be fine if someone else could give it a once over because it includes my code?

Sama34 reviewed (approved):
I have reviewed it and everything looks to be working as expected.

However, I disagree with the approach in `settings.php` described in the comment.

JordanMussi commented:
Yes I agree with @Sama34 that we should be consistent and allow through the setting changes except for the one to be rejected.

dvz commented:
It looks like upgrade script forms were only used to get around execution time limits and to introduce optional new features in the past.

This may complicate future CLI migrations and new upgrade scripts, where no manual input or output would be preferred.

https://github.com/mybb/mybb/issues/3955#issuecomment-633140416

yuliu commented:
> This may complicate future CLI migrations and new upgrade scripts, where no manual input or output would be preferred.

I can go with the _temporary_ solution for now.

I think we'll eventually end up implementing a sanity check for several important configurations like `adminemail` and admin's password/credentials etc. in the admin area. And then we can reply on it regardless of upgrade scripts altering user-data.

euantorano commented:
I'm happy with the upgrade script personally.

As far as a CLI goes, the chances are that it'll only be used to run upgrades provided after the CLI is available - I don't see how/why you'd run it before that point.

euantorano committed:
Use validate_email_format

Use `validate_email_format` core function to check for a valid email, and handle errors as handled elsewhere by unseating the setting.

euantorano committed:
Update language entry to match format

Other language entries for errors use the `.smalltext` CSS class for styling, such as `success_settings_updated_hiddencaptchaimage`.

euantorano committed:
Code style changes for upgrade script

Follow MyBB code style, and add some spacing for the sake of clarity.

dvz commented on install/resources/upgrade50.php:
```diff
@@ -32,6 +32,151 @@ function upgrade50_dbchanges()
+{
+	global $output, $db, $mybb, $cache, $errors, $checked;
+	
+	if(empty($mybb->get_input('usepresetemail')) || !in_array($mybb->input['usepresetemail'], array('current_user', 'smtp', 'custom')))
```
> Arbitrary expressions in empty are allowed in PHP 5.5 only

Sama34 commented:
Regarding the CLI topic, I think we can standardize the script later on to avoid this kind of unexpected or _complex_ changes.
